### PR TITLE
[FSDP] Consolidate FSDP state_dict offload_to_cpu settings

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86912
* #86243
* __->__ #86211

Consolidate FSDP state_dict offload_to_cpu settings. All state_dict_types now
have offload_to_cpu options.

Differential Revision: [D40065969](https://our.internmc.facebook.com/intern/diff/D40065969/)

### CODE REVIEW

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks! Do we need to update the appropriate documentation anywhere?

File: test/distributed/fsdp/test_fsdp_state_dict.py [hunk 1]
```diff
@@ -180,8 +185,16 @@ def _get_state_dict_mgr(
...
+            config = LocalStateDictConfig(
+                offload_to_cpu=state_dict_rank0_and_offload,
+            )
+        elif state_dict_type == "sharded_state_dict":
+            config = ShardedStateDictConfig(
+                offload_to_cpu=state_dict_rank0_and_offload,
+            )
         else:
-            config = None
+            raise ValueError("Unspported state_dict_type")
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_state_dict.py:197]: nit:
```suggestion
            raise ValueError("Unsupported state_dict_type")
```

File: test/distributed/fsdp/test_fsdp_state_dict.py [hunk 2]
```diff
@@ -13,7 +13,12 @@
...
 )
-from torch.distributed.fsdp import CPUOffload, FullStateDictConfig
+from torch.distributed.fsdp import (
+    CPUOffload,
+    FullStateDictConfig,
+    LocalStateDictConfig,
+    ShardedStateDictConfig,
+)
 from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
 from torch.distributed.fsdp import (
```
Comment (awgu) [inline test/distributed/fsdp/test_fsdp_state_dict.py:23]: nit: Possibly sort imports to fix the lint?

File: test/distributed/fsdp/test_fsdp_state_dict.py [hunk 3]
```diff
@@ -73,8 +78,8 @@
 
 NON_ROOT_FSDP_PREFIX = 'non_fsdp_lin'
 
-_UNFLATTENED_STATE_DICT_IMPLS = ["state_dict", "sharded_state_dict"]
-_FLATTENED_STATE_DICT_IMPLS = ["local_state_dict"]
+_UNFLATTENED_STATE_DICT_IMPLS = ["state_dict"]
+_FLATTENED_STATE_DICT_IMPLS = []
```
[the remark this answers is not in the document]
Comment (fegin) [reply test/distributed/fsdp/test_fsdp_state_dict.py:82]: oops, accidentally commit this change. Revert it.

**Pull request closed without merging**